### ISSUE TITLE
SNOW-996732: Error - cannot import name 'SnowflakeOCSPAsn1Crypto' from partially initialized module

### ISSUE THREAD

**Issue by yogevyuval** [snowflakedb/snowflake-connector-python#1840]
### Python version

Python 3.11.4 (main, Jun 13 2023, 15:34:37) [GCC 8.3.0]

### Operating system and processor architecture

Linux-5.10.198-187.748.amzn2.x86_64-x86_64-with-glibc2.28

### Installed packages

```python
aio-pika==8.3.0
aiobotocore==2.7.0
aiohttp==3.9.1
aioitertools==0.11.0
aioresponses==0.7.3
aiormq==6.6.4
aiosignal==1.3.1
alembic==1.13.0
amqp==5.2.0
asn1crypto==1.5.1
async-timeout==4.0.3
attrs==21.4.0
backcall==0.2.0
backoff==2.1.2
bcrypt==4.0.0
beautifulsoup4==4.9.3
betamax==0.8.1
bidict==0.22.1
billiard==3.6.4.0
bleach==4.1.0
blinker==1.7.0
boltons==23.1.1
boto3==1.28.23
botocore==1.31.64
build==0.10.0
cachetools==4.2.4
celery==5.2.2
certifi==2023.11.17
cffi==1.16.0
charset-normalizer==3.3.2
click==8.1.7
click-didyoumean==0.3.0
click-plugins==1.1.1
click-repl==0.3.0
comm==0.2.0
contextlib2==21.6.0
contexttimer==0.3.3
coverage==6.4.4
cryptography==41.0.7
cssselect==1.2.0
datadog==0.47.0
debugpy==1.8.0
decorator==5.1.1
distro==1.8.0
dnspython==2.4.2
elasticsearch==7.17.9
elasticsearch-dsl==7.4.1
entrypoints==0.4
eventlet==0.33.3
fastjsonschema==2.19.0
filelock==3.13.1
Flask==2.0.3
Flask-Cors==4.0.0
Flask-JWT-Extended==4.6.0
Flask-SocketIO==5.1.0
freezegun==1.1.0
frozenlist==1.4.1
funcsigs==1.0.2
greenlet==3.0.2
gunicorn==21.2.0
h11==0.14.0
html5lib==1.1
humanize==3.14.0
idna==3.6
iniconfig==2.0.0
ipykernel==6.27.1
ipython==7.34.0
itsdangerous==2.1.2
jedi==0.19.1
Jinja2==3.1.2
jmespath==1.0.1
jsonschema==4.17.3
jupyter_client==8.6.0
jupyter_core==5.5.1
kombu==5.3.4
Logbook==1.5.3
lxml==4.9.4
Mako==1.3.0
MarkupSafe==2.1.3
matplotlib-inline==0.1.6
more-itertools==8.14.0
multidict==6.0.4
nbclient==0.9.0
nbformat==5.9.2
nest-asyncio==1.5.8
numpy==1.26.2
oscrypto==1.3.0
packaging==23.2
pamqp==3.2.1
pandas==1.5.3
papermill==2.5.0
parso==0.8.3
pendulum==2.1.2
pexpect==4.9.0
pickleshare==0.7.5
pika==1.3.2
platformdirs==3.11.0
pluggy==1.3.0
prompt-toolkit==3.0.43
protobuf==4.25.1
psutil==5.9.7
psycopg2==2.9.5
ptyprocess==0.7.0
py==1.11.0
pyarrow==10.0.1
pycparser==2.21
pycryptodome==3.19.0
pycryptodomex==3.19.0
pydash==5.1.2
Pygments==2.17.2
PyJWT==2.8.0
pyOpenSSL==23.3.0
pyprinter==1.5.3
pyproject_hooks==1.0.0
pyrsistent==0.20.0
pytest==7.1.3
pytest-asyncio==0.18.3
pytest-env==0.8.1
pytest-flask==1.2.0
pytest-mock==3.8.2
pytest-rerunfailures==13.0
pytest-timeout==2.1.0
python-dateutil==2.8.2
python-engineio==4.8.0
python-socketio==5.10.0
pytz==2023.3.post1
pytzdata==2020.1
PyYAML==6.0.1
pyzmq==25.1.2
raven==6.10.0
redis==4.6.0
requests==2.28.2
retry==0.9.2
rico==1.2.2
rook==0.1.196
s3path==0.4.2
s3transfer==0.6.2
schema==0.7.5
scrapbook==0.5.0
selenium==3.141.0
sentry-sdk==1.21.0
setproctitle==1.2.2
simple-websocket==1.0.0
simplejson==3.19.2
six==1.16.0
ski==1.0.264
slackclient==2.9.4
smart-open==6.4.0
snowflake-connector-python==3.3.1
snowflake-sqlalchemy==1.5.0
sortedcontainers==2.4.0
soupsieve==2.5
SQLAlchemy==1.4.49
sqlglot==20.1.0
tenacity==8.2.3
tomli==2.0.1
tomlkit==0.12.3
toolz==0.12.0
tornado==6.4
tqdm==4.66.1
traitlets==5.14.0
typing_extensions==4.9.0
ujson==5.5.0
urllib3==1.26.18
urlquote==2.0.0
vine==5.1.0
wcwidth==0.2.12
webencodings==0.5.1
websocket-client==1.7.0
Werkzeug==2.2.3
wrapt==1.16.0
wsproto==1.2.0
XlsxWriter==1.4.5
xpath-py==0.1.2
yarl==1.9.4
```

### What did you do?

```python
We are getting intermittent errors that does not easily replicate:
cannot import name 'SnowflakeOCSPAsn1Crypto' from partially initialized module 'snowflake.connector.ocsp_asn1crypto' (most likely due to a circular import) (/usr/local/lib/python3.11/site-packages/snowflake/connector/ocsp_asn1crypto.py)

We suspect it might have something to do with the session getting some error and rolling back with an error, and then ending up in a broken state, but it's just a guess and hard to tell.

There is a previous issue about this that was already closed, apperantly the fix was to upgrade python, but this is already in new python version 3.11
https://github.com/snowflakedb/snowflake-connector-python/issues/1073
```

### What did you expect to see?

No import error should be present

**Comment by sfc-gh-mkeller** [comment]
I tried reproducing this to no avail on my Mac with `Python 3.11.6`. I installed all non-python dependencies with brew and here's the requirements file I used:
```
aio-pika==8.3.0
aiobotocore==2.7.0
aiohttp==3.9.1
aioitertools==0.11.0
aioresponses==0.7.3
aiormq==6.6.4
aiosignal==1.3.1
alembic==1.13.0
amqp==5.2.0
asn1crypto==1.5.1
async-timeout==4.0.3
attrs==21.4.0
backcall==0.2.0
backoff==2.1.2
bcrypt==4.0.0
beautifulsoup4==4.9.3
betamax==0.8.1
bidict==0.22.1
billiard==3.6.4.0
bleach==4.1.0
blinker==1.7.0
boltons==23.1.1
boto3==1.28.23
botocore==1.31.64
build==0.10.0
cachetools==4.2.4
capybara-py==0.1.6
celery==5.2.2
certifi==2023.11.17
cffi==1.16.0
charset-normalizer==3.3.2
click==8.1.7
click-didyoumean==0.3.0
click-plugins==1.1.1
click-repl==0.3.0
comm==0.2.0
contextlib2==21.6.0
contexttimer==0.3.3
coverage==6.4.4
cryptography==41.0.7
cssselect==1.2.0
datadog==0.47.0
debugpy==1.8.0
decorator==5.1.1
distro==1.8.0
dnspython==2.4.2
elasticsearch==7.17.9
elasticsearch-dsl==7.4.1
entrypoints==0.4
eventlet==0.33.3
fastjsonschema==2.19.0
filelock==3.13.1
Flask==2.0.3
Flask-Cors==4.0.0
Flask-JWT-Extended==4.6.0
Flask-SocketIO==5.1.0
freezegun==1.1.0
frozenlist==1.4.1
funcsigs==1.0.2
greenlet==3.0.2
gunicorn==21.2.0
h11==0.14.0
html5lib==1.1
humanize==3.14.0
idna==3.6
iniconfig==2.0.0
ipykernel==6.27.1
ipython==7.34.0
itsdangerous==2.1.2
jedi==0.19.1
Jinja2==3.1.2
jmespath==1.0.1
jsonschema==4.17.3
jupyter_client==8.6.0
jupyter_core==5.5.1
kombu==5.3.4
Logbook==1.5.3
lxml==4.9.4
Mako==1.3.0
MarkupSafe==2.1.3
matplotlib-inline==0.1.6
more-itertools==8.14.0
multidict==6.0.4
nbclient==0.9.0
nbformat==5.9.2
nest-asyncio==1.5.8
numpy==1.26.2
oscrypto==1.3.0
packaging==23.2
pamqp==3.2.1
pandas==1.5.3
papermill==2.5.0
parso==0.8.3
pendulum==2.1.2
pexpect==4.9.0
pickleshare==0.7.5
pika==1.3.2
platformdirs==3.11.0
pluggy==1.3.0
prompt-toolkit==3.0.43
protobuf==4.25.1
psutil==5.9.7
psycopg2==2.9.5
ptyprocess==0.7.0
py==1.11.0
pyarrow==10.0.1
pycparser==2.21
pycryptodome==3.19.0
pycryptodomex==3.19.0
pydash==5.1.2
Pygments==2.17.2
PyJWT==2.8.0
pyOpenSSL==23.3.0
pyprinter==1.5.3
pyproject_hooks==1.0.0
pyrsistent==0.20.0
pytest==7.1.3
pytest-asyncio==0.18.3
pytest-env==0.8.1
pytest-flask==1.2.0
pytest-mock==3.8.2
pytest-rerunfailures==13.0
pytest-timeout==2.1.0
python-dateutil==2.8.2
python-engineio==4.8.0
python-socketio==5.10.0
pytz==2023.3.post1
pytzdata==2020.1
PyYAML==6.0.1
pyzmq==25.1.2
raven==6.10.0
redis==4.6.0
requests==2.28.2
retry==0.9.2
rico==0.4.0
rook==0.1.196
s3path==0.4.2
s3transfer==0.6.2
schema==0.7.5
scrapbook==0.5.0
selenium==3.141.0
sentry-sdk==1.21.0
setproctitle==1.2.2
simple-websocket==1.0.0
simplejson==3.19.2
six==1.16.0
ski==0.1.12
slackclient==2.9.4
smart-open==6.4.0
snowflake-connector-python==3.3.1
snowflake-sqlalchemy==1.5.0
sortedcontainers==2.4.0
soupsieve==2.5
SQLAlchemy==1.4.49
sqlglot==20.1.0
tenacity==8.2.3
tomli==2.0.1
tomlkit==0.12.3
toolz==0.12.0
tornado==6.4
tqdm==4.66.1
traitlets==5.14.0
typing_extensions==4.9.0
ujson==5.5.0
urllib3==1.26.18
urlquote==2.0.0
vine==5.1.0
wcwidth==0.2.12
webencodings==0.5.1
websocket-client==1.7.0
Werkzeug==2.2.3
wrapt==1.16.0
wsproto==1.2.0
XlsxWriter==1.4.5
xpath-py==0.1.2
yarl==1.9.4
```
@yogevyuval some of your packages' version doesn't seem to be available on PyPi. For example `capybara-py==0.1.10` doesn't seem to be available on https://pypi.org/project/capybara/#history 
Could you please advise on where did you get these versions from?

**Comment by yogevyuval** [comment]
> I tried reproducing this to no avail on my Mac with `Python 3.11.6`. I installed all non-python dependencies with brew and here's the requirements file I used:
> 
> ```
> aio-pika==8.3.0
> aiobotocore==2.7.0
> aiohttp==3.9.1
> aioitertools==0.11.0
> aioresponses==0.7.3
> aiormq==6.6.4
> aiosignal==1.3.1
> alembic==1.13.0
> amqp==5.2.0
> asn1crypto==1.5.1
> async-timeout==4.0.3
> attrs==21.4.0
> backcall==0.2.0
> backoff==2.1.2
> bcrypt==4.0.0
> beautifulsoup4==4.9.3
> betamax==0.8.1
> bidict==0.22.1
> billiard==3.6.4.0
> bleach==4.1.0
> blinker==1.7.0
> boltons==23.1.1
> boto3==1.28.23
> botocore==1.31.64
> build==0.10.0
> cachetools==4.2.4
> capybara-py==0.1.6
> celery==5.2.2
> certifi==2023.11.17
> cffi==1.16.0
> charset-normalizer==3.3.2
> click==8.1.7
> click-didyoumean==0.3.0
> click-plugins==1.1.1
> click-repl==0.3.0
> comm==0.2.0
> contextlib2==21.6.0
> contexttimer==0.3.3
> coverage==6.4.4
> cryptography==41.0.7
> cssselect==1.2.0
> datadog==0.47.0
> debugpy==1.8.0
> decorator==5.1.1
> distro==1.8.0
> dnspython==2.4.2
> elasticsearch==7.17.9
> elasticsearch-dsl==7.4.1
> entrypoints==0.4
> eventlet==0.33.3
> fastjsonschema==2.19.0
> filelock==3.13.1
> Flask==2.0.3
> Flask-Cors==4.0.0
> Flask-JWT-Extended==4.6.0
> Flask-SocketIO==5.1.0
> freezegun==1.1.0
> frozenlist==1.4.1
> funcsigs==1.0.2
> greenlet==3.0.2
> gunicorn==21.2.0
> h11==0.14.0
> html5lib==1.1
> humanize==3.14.0
> idna==3.6
> iniconfig==2.0.0
> ipykernel==6.27.1
> ipython==7.34.0
> itsdangerous==2.1.2
> jedi==0.19.1
> Jinja2==3.1.2
> jmespath==1.0.1
> jsonschema==4.17.3
> jupyter_client==8.6.0
> jupyter_core==5.5.1
> kombu==5.3.4
> Logbook==1.5.3
> lxml==4.9.4
> Mako==1.3.0
> MarkupSafe==2.1.3
> matplotlib-inline==0.1.6
> more-itertools==8.14.0
> multidict==6.0.4
> nbclient==0.9.0
> nbformat==5.9.2
> nest-asyncio==1.5.8
> numpy==1.26.2
> oscrypto==1.3.0
> packaging==23.2
> pamqp==3.2.1
> pandas==1.5.3
> papermill==2.5.0
> parso==0.8.3
> pendulum==2.1.2
> pexpect==4.9.0
> pickleshare==0.7.5
> pika==1.3.2
> platformdirs==3.11.0
> pluggy==1.3.0
> prompt-toolkit==3.0.43
> protobuf==4.25.1
> psutil==5.9.7
> psycopg2==2.9.5
> ptyprocess==0.7.0
> py==1.11.0
> pyarrow==10.0.1
> pycparser==2.21
> pycryptodome==3.19.0
> pycryptodomex==3.19.0
> pydash==5.1.2
> Pygments==2.17.2
> PyJWT==2.8.0
> pyOpenSSL==23.3.0
> pyprinter==1.5.3
> pyproject_hooks==1.0.0
> pyrsistent==0.20.0
> pytest==7.1.3
> pytest-asyncio==0.18.3
> pytest-env==0.8.1
> pytest-flask==1.2.0
> pytest-mock==3.8.2
> pytest-rerunfailures==13.0
> pytest-timeout==2.1.0
> python-dateutil==2.8.2
> python-engineio==4.8.0
> python-socketio==5.10.0
> pytz==2023.3.post1
> pytzdata==2020.1
> PyYAML==6.0.1
> pyzmq==25.1.2
> raven==6.10.0
> redis==4.6.0
> requests==2.28.2
> retry==0.9.2
> rico==0.4.0
> rook==0.1.196
> s3path==0.4.2
> s3transfer==0.6.2
> schema==0.7.5
> scrapbook==0.5.0
> selenium==3.141.0
> sentry-sdk==1.21.0
> setproctitle==1.2.2
> simple-websocket==1.0.0
> simplejson==3.19.2
> six==1.16.0
> ski==0.1.12
> slackclient==2.9.4
> smart-open==6.4.0
> snowflake-connector-python==3.3.1
> snowflake-sqlalchemy==1.5.0
> sortedcontainers==2.4.0
> soupsieve==2.5
> SQLAlchemy==1.4.49
> sqlglot==20.1.0
> tenacity==8.2.3
> tomli==2.0.1
> tomlkit==0.12.3
> toolz==0.12.0
> tornado==6.4
> tqdm==4.66.1
> traitlets==5.14.0
> typing_extensions==4.9.0
> ujson==5.5.0
> urllib3==1.26.18
> urlquote==2.0.0
> vine==5.1.0
> wcwidth==0.2.12
> webencodings==0.5.1
> websocket-client==1.7.0
> Werkzeug==2.2.3
> wrapt==1.16.0
> wsproto==1.2.0
> XlsxWriter==1.4.5
> xpath-py==0.1.2
> yarl==1.9.4
> ```
> 
> @yogevyuval some of your packages' version doesn't seem to be available on PyPi. For example `capybara-py==0.1.10` doesn't seem to be available on https://pypi.org/project/capybara/#history Could you please advise on where did you get these versions from?

As I said, unfortunalty it's not reproducible easily,  it happens intermittently, but it defintely happens. A google search will show tens of peoples reporting it - but it's more complex than just importing and getting that error, it's a specific flow related issue. About the package, I removed it from the original message, it's not relevant.

**Comment by yogevyuval** [comment]
@sfc-gh-mkeller Here is some more context from Sentry:
<img width="1110" alt="image" src="https://github.com/snowflakedb/snowflake-connector-python/assets/962869/1d30def1-ec05-422b-8080-a874458e476d">

**Comment by yogevyuval** [comment]
> @sfc-gh-mkeller Here is some more context from Sentry: <img alt="image" width="1110" src="https://private-user-images.githubusercontent.com/962869/292573815-1d30def1-ec05-422b-8080-a874458e476d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ0OTEyNjMsIm5iZiI6MTcwNDQ5MDk2MywicGF0aCI6Ii85NjI4NjkvMjkyNTczODE1LTFkMzBkZWYxLWVjMDUtNDIyYi04MDgwLWE4NzQ0NThlNDc2ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwNVQyMTQyNDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MTgwY2QwNmM3ZjhjN2JhOWM3Y2VlNWI4MWFkMjU1ZDA0MDcwNzg3NjMzYTA2OTZkN2YxZDk0YjEyNjdjMDlkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.fUpIRtP8Pn_JukAeN7_cMDJjDk9CL7AI3-FX-3CCpyI">

@sfc-gh-mkeller Is this helpful? Do you need any more information for this? (not sure i have any more haha)

**Comment by sfc-gh-aling** [comment]
hey @yogevyuval , I have some updates for you.

we recently got another report for this circular import error, and by looking at the log of the other report, we find the circular import might by caused by `__del__` of the connection object being called while the connection is being initialized.

this explains why it's intermittent because `__del__` is called randomly by python interpreter.
we're still trying to understand and reproduce the error, in the meanwhile we are seeking to refactor the code to avoid `__del__` to be called while a connection object is being constructed or avoid performing http request in the `__del__` method.

**Comment by sfc-gh-dszmolka** [comment]
PR is now merged and will be part of the next driver version, expected to be released towards end of March

**Comment by sfc-gh-dszmolka** [comment]
fix released with v3.8.0, thank you for bearing with us !
